### ISSUE TITLE
feature request: trigger an SSE start with the hx-trigger

### ISSUE THREAD

**Issue by GrimalDev** [bigskysoftware/htmx-extensions#27]
Hello,

I found in the documentation that it is possible to trigger http callbacks on a specific sse event.
But I do not find anything (and don't seem to be able to get it to work) about using the hx-trigger tag to start sse requests on a user event (like a click to start the server sent events).

forgive me if I missed something in the documentation, but it isn't clear to me why in all my tests, my sse event are just triggered out of the box on page load.

Thank a lot for your help all.

**Comment by Telroshan** [comment]
Hello, I'm not sure to understand the issue, sorry if that's off the mark;
> using the hx-trigger tag to start sse requests on a user event (like a click to start the server sent events)

It looks like you want to send messages  **from** the client **to** the server, but please note [SSE (**Server** sent events)](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events), is unidirectional, and can only happen from the server to the client.
If you want bidirectional communication, you'll probably be more interested in the [websockets extension](https://htmx.org/extensions/web-sockets/), which will allow you to send messages from your client to your server.


> it isn't clear to me why in all my tests, my sse event are just triggered out of the box on page load.

That sounds weird, SSE events shouldn't trigger unless your server sends those very SSE events in the first place. Would you mind sharing some code to investigate on what's happening here?

**Comment by Telroshan** [comment]
Ooh ok I get it, you want to initiate the SSE connection upon user interaction.

Is there any reason not to have it initialize the SSE connection automatically on page load though? You could just have it active & idle, until you actually send events from your server.

Other than that, there's no current built-in way to do this that I know of, but you could achieve this by
- If your click makes an htmx request, your endpoint could include a `<div sse-connect="/url">` element in its response, which would initiate the SSE connection once swapped in by htmx. If it's a pure client side interaction though, and your button doesn't fetch anything from the server, ofc you can forget this one.
- With a little custom JS snippet, create an element on the go on click, add the `sse-connect` attribute to it, insert it in the DOM wherever you'd like, and call `htmx.process` on it to have it initialize the SSE connection

The same ideas apply if you want to instead act on the `hx-ext` attribute, which you could use to say enable the SSE extension altogether on click (note this will affect all events, not just the connection initialization).

Basically, `sse-connect` is an attribute that instructs to initiate a SSE connection to the given endpoint, so if you don't want the SSE connection on load, remove this attribute, or avoid having a `hx-ext="sse"` parent element, to later on introduce any of those 2 when you want it to happen

**Comment by GrimalDev** [comment]
> Ooh ok I get it, you want to initiate the SSE connection upon user interaction.
> 
> Is there any reason not to have it initialize the SSE connection automatically on page load though? You could just have it active & idle, until you actually send events from your server.
> 
> Other than that, there's no current built-in way to do this that I know of, but you could achieve this by
> 
> * If your click makes an htmx request, your endpoint could include a `<div sse-connect="/url">` element in its response, which would initiate the SSE connection once swapped in by htmx. If it's a pure client side interaction though, and your button doesn't fetch anything from the server, ofc you can forget this one.
> * With a little custom JS snippet, add the `sse-connect` attribute on the go, and call `htmx.process` on the element to have it initialize the SSE connection
> 
> The same ideas apply if you want to instead act on the `hx-ext` attribute, which you could use to say enable the SSE extension altogether on click (note this will affect all events, not just the connection initialization).
> 
> Basically, `sse-connect` is an attribute that instructs to initiate a SSE connection to the given endpoint, so if you don't want the SSE connection on load, remove this attribute, or avoid having a `hx-ext="sse"` parent element, to later on introduce any of those 2 when you want it to happen

Thank you for your detailed explanation, I will do something like adding the sse-connect tag on button click. But still, It seems like a lot of effort for something that could be a feature. Of course, it would take time, maybe i'll look at making a code change proposition for this edge case. But thank you for your time.

**Comment by trafflux** [comment]
I had this same issue and I do not like the solution proposed. sse_connect should be tied to an event to be consistent.

**Comment by Telroshan** [comment]
Seems I somehow forgot to post a message when adding the "enhancement" label back then, but sure I don't see any reason to oppose to making the extension more flexible & configurable.
If any of you is interested in diving into this and suggesting an implementation with a PR, please feel free to do so!

As a side note, extensions are way more open when it comes to changes than the core library, partly due to the fact that each extension is a specific feature & has a relatively small code, see how this repo's [contribution guidelines](https://github.com/bigskysoftware/htmx-extensions/blob/main/CONTRIBUTING.md) do not require a maintainer's approval to submit an enhancement PR 😉

**Comment by y2kbugger** [comment]
I have forked the sse extension (not open source) to add exactly this, as well as hx-indicator.

The aggressive auto-connect and reconnect is only imagining that you have an endpoint that sends events that are not anchored in time.

For example, lets say you have a slow operation that does ETL from 5 tables, you could put it on a button, and then stream progress updates.

``` html
<button
  sse-connect="/refresh_erp_transactions"
  sse-swap="message"
  sse-close="done"
  hx-trigger="click" 
  hx-indicator="this"
  hx-swap="innerHTML"
>Import Data</button>
```

This allows for more locality of behavior for both frontend and backend.

I think right now the design pressures you keep a single, eternal, SSE connection open and then all long running operation as well as other events channel though it like a global event queue. (please correct if wrong

**Comment by nicolasfranck** [comment]
Another reason for changing the default trigger (which is `hx-trigger="load"` to call sse endpoint on page load),
is to delay that event. 

Imagine that this extension is installed in the `<body>` on every page you have,
and that a user starts navigating fast between pages. If the background call to the sse endpoint has not finished
yet before the page is changed, the browser waits for 1 minute until that is done, and you see the page loading icon appearing
for no apparent reason (end users are not aware that they should wait for that background call).

I have noticed this behaviour, and indeed solution 1 is to track all sse connections and abort/close them on page unload (which works). But postponing the connection can also be helpful. If you wait for 2 seconds for example, odds are greater that the end user will stay to receive the messages sent by sse.
